app/services/socket.js

```diff
@@ -0,0 +1,52 @@
+import PhoenixSocket from 'phoenix/services/phoenix-socket';
+import { get } from '@ember/object';
+import { inject as service } from '@ember/service';
+import ENV from 'code-corps-ember/config/environment';
+
+export default PhoenixSocket.extend({
+  session: service(),
+
+  init() {
+    let session = get(this, 'session');
+    session.on('authenticationSucceeded', () => this.toggleConnection());
+    session.on('invalidationSucceeded', () => this.toggleConnection());
+  },
+
+  connect() {
+    let socket = get(this, 'socket');
+    let token = get(this, 'session.session.authenticated.token');
+    let params = token ? { token } : {};
+
+    // Only create a socket when:
+    // - no socket exists
+    // - we're exchanging an unauthenticated socket for an authenticated one
+    if (!socket) {
+      this._super(ENV.SOCKETS_BASE_URL, { params });
+    } else if (socket && token) {
+      socket.disconnect(() => this._super(ENV.SOCKETS_BASE_URL, { params }));
+    }
+  },
+
+  disconnect(callback) {
+    let socket = get(this, 'socket');
+    if (socket) {
+      socket.disconnect(callback);
+    } else {
+      return callback && callback();
+    }
+  },
+
+  joinChannel(channelName) {
```

Comment:
I wonder if some doc blocks for a few of these methods might be helpful...

Reply:
These are really just wrappers around `connect` and `disconnect` for `PhoenixSocket` from `ember-phoenix`.